lib/paloaltofw.py

```diff
@@ -0,0 +1,583 @@
+    """Representation of an individual term.
+
+       This is mostly useful for the __str__() method.
+
+       Args:
+         obj: a policy.Term object
+         term_type: type of filter to generate, e.g. inet or inet6
+         filter_options: list of remaining target options (zones)
+    """
+
+    ACTIONS = {'accept': 'allow',
+               'deny': 'deny',
+               'reject': 'reject',
+               'count': 'count',
+               'log': 'log'}
+
+    def __init__(self, term, term_type, zones):
+        self.term = term
+        self.term_type = term_type
+        self.from_zone = zones[1]
+        self.to_zone = zones[3]
+        self.extra_actions = []
+
+    def __str__(self):
+        """Render config output from this term object."""
+        # Verify platform specific terms. Skip whole term if platform does not
+        # match.
+        # Nothing here for now
+
+    def _Group(self, group):
+        """If 1 item return it, else return [ item1 item2 ].
+
+        Args:
+          group: a list.  could be a list of strings (protocols) or a list of
+                 tuples (ports)
+
+        Returns:
+          rval: a string surrounded by '[' and '];' if len(group) > 1
+                or with just ';' appended if len(group) == 1
+        """
+
+        def _FormattedGroup(el):
+            """Return the actual formatting of an individual element.
+
+            Args:
+              el: either a string (protocol) or a tuple (ports)
+
+            Returns:
+              string: either the lower()'ed string or the ports, hyphenated
+                      if they're a range, or by itself if it's not.
+            """
+            if isinstance(el, str):
+                return el.lower()
+            elif isinstance(el, int):
+                return str(el)
+            # type is a tuple below here
+            elif el[0] == el[1]:
+                return '%d' % el[0]
+            else:
+                return '%d-%d' % (el[0], el[1])
+
+        if len(group) > 1:
+            rval = '[ ' + ' '.join([_FormattedGroup(x) for x in group]) + ' ];'
+        else:
+            rval = _FormattedGroup(group[0]) + ';'
+        return rval
+
+
+class Service():
+
+    service_map = {}
+
+    def __init__(self, ports, service_name, protocol):  # ports is a tuple of ports
+        if (ports, protocol) in self.service_map:
+            raise PaloAltoFWDuplicateServiceError(
+            'You have a duplicate service. A service already exists on port(s): %s'
+                                                  % str(ports))
+
+        final_service_name = 'service-' + service_name + '-' + protocol
+        if len(final_service_name.decode('utf-8')) > 63:
+            raise PaloAltoFWTooLongName(
+            'Service name must be 63 characters max: %s'
+                                        % str(final_service_name))
+        self.service_map[(ports, protocol)] = {'name': final_service_name}
+
+
+class Rule():
```

Comment:
Can we explore why this was chosen and if this could be rolled up into a term? I don't understand why this is being used this way and I don't like having code integrated that no one understands. I will expect a docstring for Rule explaining what it is and why it is used instead of the term object.